File: vtp/src/main/java/cn/vove7/vtp/system/DeviceInfo.java
package cn.vove7.vtp.system;

import android.content.Context;
import android.os.Build;

import java.util.Locale;

/**
 * @author Vove
 * <p>
 * 2018/8/15
 */
public class DeviceInfo {
    public String IMEI;
    public ScreenInfo screenInfo;
    public String manufacturerName;
    public String productName;
    public String brandName;
    public String model;
    public String boardName;
    public String deviceName;
    public String serial;
    public int sdkInt;
    public String androidVersion;
    public String language;

    private DeviceInfo(String IMEI, ScreenInfo screenInfo, String manufacturerName,
                       String productName, String brandName, String model, String boardName,
                       String deviceName, String serial, int sdkInt, String androidVersion,
                       String language) {
        this.IMEI = IMEI;
        this.screenInfo = screenInfo;
        this.manufacturerName = manufacturerName;
        this.productName = productName;
        this.brandName = brandName;
        this.model = model;
        this.boardName = boardName;
        this.deviceName = deviceName;
        this.serial = serial;
        this.sdkInt = sdkInt;
        this.androidVersion = androidVersion;
        this.language = language;
    }

    private static DeviceInfo instance;

    public static DeviceInfo getInfo(Context context) {
        return getInfo(context, false);
    }

    public static DeviceInfo getInfo(Context context, boolean needImei) {
        if (instance == null) {
            init(context, needImei);
        } else {
            instance.language = Locale.getDefault().getLanguage();
        }
        return instance;

    }

    private static void init(Context context, boolean b) {
        instance = new DeviceInfo(
                b ? SystemHelper.INSTANCE.getIMEI(context) : null,
                SystemHelper.INSTANCE.getScreenInfo(context),
                Build.MANUFACTURER,
                Build.PRODUCT,
                Build.BRAND,
                Build.MODEL,
                Build.BOARD,
                Build.DEVICE,
                Build.SERIAL,
                Build.VERSION.SDK_INT,
                Build.VERSION.RELEASE,
                Locale.getDefault().getLanguage()
        );
    }
}

File: vtp/src/main/java/cn/vove7/vtp/system/DeviceInfo.kt
package cn.vove7.vtp.system

import android.content.Context
import android.os.Build
import java.util.*

/**
 * # DeviceInfo
 *
 * @author Vove
 * 2018/8/15
 */
data class DeviceInfo(
        var IMEI: String? = null,//IMEI
        var screenSInfo: ScreenInfo?,//屏幕信息
        var manufacturerName: String,//厂商名
        var productName: String,//产品名
        var brandName: String,//品牌
        var model: String,//型号
        var boardName: String,//主板名
        var deviceName: String,//设备名
        var serial: String,//序列号
        var sdkInt: Int,//Sdk版本
        var androidVersion: String,//Android版本

        var language: String//当前语言
) {
    companion object {
        private var instance: DeviceInfo? = null
        fun getInfo(context: Context, needImei: Boolean = false): DeviceInfo {
            if (instance == null) {
                init(context, needImei)
            } else {
                instance!!.language = Locale.getDefault().language
            }
            return instance!!
        }

        private fun init(context: Context, b: Boolean) {
            instance = DeviceInfo(
                    if (b) SystemHelper.getIMEI(context) else null,
                    SystemHelper.getScreenInfo(context),
                    Build.MANUFACTURER,
                    Build.PRODUCT,
                    Build.BRAND,
                    Build.MODEL,
                    Build.BOARD,
                    Build.DEVICE,
                    Build.SERIAL,
                    Build.VERSION.SDK_INT,
                    Build.VERSION.RELEASE,
                    Locale.getDefault().language
            )
        }
    }
}
